pkgs/development/libraries/gsm/default.nix

```diff
@@ -1,41 +1,51 @@
 { stdenv, fetchurl }:
 
-with stdenv.lib;
+let
+  inherit (stdenv) isCygwin isDarwin;
+  inherit (stdenv.lib) optional;
+in
+
 stdenv.mkDerivation rec {
   name = "gsm-${version}";
-  version = "1.0.13";
+  version = "1.0.14";
 
   src = fetchurl {
     url = "http://www.quut.com/gsm/${name}.tar.gz";
-    sha256 = "1bcjl2h60gvr1dc5a963h3vnz9zl6n8qrfa3qmb2x3229lj1iiaj";
+    sha256 = "0b1mx69jq88wva3wk0hi6fcl5a52qhnq2f9p3f3jdh5k61ma252q";
   };
 
-  patchPhase = ''
-    # Fix include directory
-    sed -e 's,$(GSM_INSTALL_ROOT)/inc,$(GSM_INSTALL_ROOT)/include/gsm,' -i Makefile
-
-    makeFlags="$makeFlags INSTALL_ROOT=$out"
-
-    # Build shared library instead of static
-    sed -e 's,-c -O2 -DNeedFunctionPrototypes=1,-c -O2 -fPIC -DNeedFunctionPrototypes=1,' -i Makefile
-    sed -e 's,libgsm.a,libgsm.so,' -i Makefile
-    sed -e 's/$(AR) $(ARFLAGS) $(LIBGSM) $(GSM_OBJECTS)/$(LD) -shared -Wl,-soname,libgsm.so -o $(LIBGSM) $(GSM_OBJECTS) -lc/' -i Makefile
-    sed -e 's,$(RANLIB) $(LIBGSM),,' -i Makefile
-  '';
+  patchPhase = [
+    # Fix prefix & include directories
+    ''
+      sed -e 's,$(GSM_INSTALL_ROOT)/inc,$(GSM_INSTALL_ROOT)/include/gsm,' -i Makefile
+      makeFlags="$makeFlags INSTALL_ROOT=$out"
+    ''
+  ] # Build ELF shared library instead of static
+    ++ optional (!isCygwin && !isDarwin) ''
+      sed -e 's,-c -O2 -DNeedFunctionPrototypes=1,-c -O2 -fPIC -DNeedFunctionPrototypes=1,' -i Makefile
+      sed -e 's,libgsm.a,libgsm.so,' -i Makefile
+      sed -e 's/$(AR) $(ARFLAGS) $(LIBGSM) $(GSM_OBJECTS)/$(LD) -shared -Wl,-soname,libgsm.so -o $(LIBGSM) $(GSM_OBJECTS) -lc/' -i Makefile
+      sed -e 's,$(RANLIB) $(LIBGSM),,' -i Makefile
+    ''
+    # Build dylib on Darwin
+    ++ optional isDarwin ''
+      sed -e 's,-c -O2 -DNeedFunctionPrototypes=1,-c -O2 -fPIC -DNeedFunctionPrototypes=1,' -i Makefile
+      sed -e 's,libgsm.a,libgsm.dylib,' -i Makefile
+      sed -e 's/$(AR) $(ARFLAGS) $(LIBGSM) $(GSM_OBJECTS)/$(LD) -o $(LIBGSM) -dynamiclib -install_name $(LIBGSM) $(GSM_OBJECTS) -lc/' -i Makefile
```

Comment:
Change the install name to `$(INSTALL_ROOT)/$(LIBGSM)`. Don't forget to change the set the separator character to something other than /. Using an absolute path is best for dylibs in nix, it helps avoid a ton of bugs.